crates/circuit/src/symbol_expr.rs

```diff
@@ -0,0 +1,3234 @@
+                            op: UnaryOp::Neg,
+                            expr: Box::new(expr_d),
+                        },
+                        UnaryOp::Sin => {
+                            let lhs = SymbolExpr::Unary {
+                                op: UnaryOp::Cos,
+                                expr: Box::new(expr.as_ref().clone()),
+                            };
+                            lhs * expr_d
+                        }
+                        UnaryOp::Asin => {
+                            let d = &SymbolExpr::Value(Value::Real(1.0))
+                                - &(expr.as_ref() * expr.as_ref());
+                            let lhs = match d {
+                                SymbolExpr::Value(v) => SymbolExpr::Value(v.sqrt()),
+                                _ => _pow(d, SymbolExpr::Value(Value::Real(0.5))),
+                            };
+                            &lhs * &expr_d
+                        }
+                        UnaryOp::Cos => {
+                            let lhs = SymbolExpr::Unary {
+                                op: UnaryOp::Sin,
+                                expr: Box::new(expr.as_ref().clone()),
+                            };
+                            &-&lhs * &expr_d
+                        }
+                        UnaryOp::Acos => {
+                            let d = &SymbolExpr::Value(Value::Real(1.0))
+                                - &(expr.as_ref() * expr.as_ref());
+                            let lhs = match d {
+                                SymbolExpr::Value(v) => SymbolExpr::Value(v.sqrt()),
+                                _ => _pow(d, SymbolExpr::Value(Value::Real(0.5))),
+                            };
+                            &-&lhs * &expr_d
+                        }
+                        UnaryOp::Tan => {
+                            let d = SymbolExpr::Unary {
+                                op: UnaryOp::Cos,
+                                expr: Box::new(expr.as_ref().clone()),
+                            };
+                            &(&expr_d / &d) / &d
+                        }
+                        UnaryOp::Atan => {
+                            let d = &SymbolExpr::Value(Value::Real(1.0))
+                                + &(expr.as_ref() * expr.as_ref());
+                            &expr_d / &d
+                        }
+                        UnaryOp::Exp => {
+                            &SymbolExpr::Unary {
+                                op: UnaryOp::Exp,
+                                expr: Box::new(expr.as_ref().clone()),
+                            } * &expr_d
+                        }
+                        UnaryOp::Log => &expr_d / expr.as_ref(),
+                        UnaryOp::Sign => SymbolExpr::Unary {
+                            op: UnaryOp::Sign,
+                            expr: Box::new(expr_d),
+                        },
+                        UnaryOp::Conj => SymbolExpr::Unary {
```

Comment:
I don't think the conjugation is differentiable for complex arguments. 

But there's no way to know whether the parameter represents a real number, right? In that case the correct thing would be to raise an error here. But I'm wondering if doing that breaks some implicit workflows, e.g. if someone conjugates a Hamiltonian and then tries to compute the derivative of the coefficients. Maybe we can try and raise an error and see if any of the tests fail.

Reply:
So I also raise error for conjugate in [7116e74](https://github.com/Qiskit/qiskit/pull/13278/commits/7116e74d64ecebb239e61b0e5c64de520740c051)